File: Python3/242. Valid Anagram.py
class Solution:
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        # charDict = {} # 字典检查两个字符串 # 96ms
        # for char in s:
        #     charDict[char] = charDict.get(char, 0) + 1
        # for char in t:
        #     if char not in charDict:
        #         return False
        #     else:
        #         charDict[char] -= 1
        # for value in charDict.values():
        #     if value:
        #         return False
        # return True
        
        # return sorted(s) == sorted(t) # 80ms
        
        # 稍微修改一下字典方法 # 76ms
        if len(s) != len(t): return False
        dict1 = collections.Counter(s)
        dict2 = collections.Counter(t)
        dict1.subtract(dict2)
        for value in dict1.values():
            if value:
                return False
        return True

File: Golang/242. Valid Anagram.go
// func isAnagram(s string, t string) bool {
//     charDict := make(map[int]int, len(s))
//     for _, char := range s {
//         charDict[int(char) - 'a']++
//     }
//     for _, char := range t {
//         charDict[int(char) - 'a']--
//     }
//     for key := range charDict {
//         if charDict[key] != 0 {
//             return false
//         }
//     }
//     return true
// }

func isAnagram(s string, t string) bool {
    if len(s) != len(t) {
        return false
    } else if len(s) == 0 {
        return true
    } else {
        charDict := [26]int{}
        for i := range s {
            charDict[int(s[i]) - 'a']++
            charDict[int(t[i]) - 'a']--
        }
        for _, count := range charDict {
            if count != 0 {
                return false
            }
        }
        return true
    }
}
